### PR TITLE
fix(graph): stabilize windows rustc response handling

### PR DESCRIPTION
**Pull request by pepicrft**
## What changed

`rustc-response` argument files now write each argument exactly as the graph action declared it. The Windows-specific escaping layer was removed because normal `rustc @path` response files should be owned by the compiler-facing materializer, not by shell quoting rules.

Rust feature flags now lower from `features = ["default"]` to compiler arguments such as `--cfg=feature="default"`, with Windows placing those arguments in a generated response file.

Windows host tool lookup now skips extensionless `PATHEXT` lookups when the caller asks for a bare tool name such as `rustc` or `cargo`. This prevents the graph analyzer from returning mise shell shims that Windows cannot spawn directly through `Command`.

Build-script crates no longer invoke native Windows `rustc` through the Unix shell wrapper. On Windows, the Rust prelude now generates a small PowerShell wrapper action that reads build-script stdout, appends `cargo:rustc-*` directives to an argument array, and invokes `rustc` directly.

`once-frontend` now declares the embedded `prelude/` tree as a Cargo rebuild input. The walk is sorted, follows directory symlinks only inside the canonical prelude root, and tracks visited directories so traversal is deterministic and bounded.

The pull request workflow includes a Windows release graph build check. It vendors the Rust dependency graph, writes the release graph manifests, and builds the `x86_64-pc-windows-msvc` Once binary through Once without packaging or publishing release assets.

## Why

The `Release` workflow on `main` is failing for the Windows package jobs while compiling Rust dependencies through Once's graph actions. The failing command reaches `rustc` with incorrectly encoded feature configuration values.

The new pull request validation also exposed two adjacent Windows graph issues: `host_which("rustc")` could resolve to an extensionless mise shim, and crates with build scripts routed native Windows `rustc` through a Unix shell wrapper.

A later validation run showed that source bootstrapping could still reuse cached `once-frontend` output after Starlark prelude changes, so the Windows job could run a stale embedded prelude even when the branch contained the corrected feature lowering.

## Root cause

The original response-file materializer treated `rustc` response files like shell-escaped command lines on Windows. That wrote backslash-escaped quotes, so `rustc` received backslash tokens and rejected the configuration argument.

After removing that escaping, the Windows release graph check showed the generated response files were correct, for example `--cfg=feature="default"`. The remaining failure only appeared on crates with build scripts. Those actions wrapped `rustc` through `sh` so Once could append `cargo:rustc-*` lines from build-script stdout, and that Windows shell path changed how the native compiler consumed the response file.

The host lookup helper also included the extensionless tool name before checking `PATHEXT` candidates on Windows. That let shell-specific shims shadow executable candidates such as `rustc.exe`.

Finally, `once-frontend` embeds the Starlark prelude at compile time, but Cargo did not know that `prelude/` was an input to the package build. With restored Rust cache outputs, a source bootstrap could reuse a stale `once-frontend` artifact after only prelude files changed.

## Approach

Move response-file responsibility to the Rust materializer: `cmd_args(format = "rustc-response")` validates that arguments are one line each, then writes them verbatim.

Move feature-flag responsibility into the Rust prelude: rule authors keep passing `features = ["default"]`, and Once lowers that to compiler-ready `--cfg=feature="default"` arguments before deciding whether to place them inline or in a response file.

Keep build-script stdout handling in the Rust target kind, but avoid the Unix shell when the host is Windows. The generated PowerShell wrapper receives the original argument vector positionally, reads the build-script stdout file, appends compiler directives as array elements, and launches `rustc` without routing the response file through Git Bash.

Make host lookup return spawnable Windows tools by honoring `PATHEXT` candidates for bare tool names and skipping the extensionless path. Target kinds can keep asking for `host_which("rustc")` and get a command the graph can execute directly.

Make the embedded prelude a declared Cargo rebuild input through a small build script. This keeps future prelude-only changes from relying on incidental Rust source changes to invalidate cached bootstrap binaries.

Add a Windows release graph build to pull request validation so this failure mode is caught before merge without invoking the release packaging task.

## Impact

Windows release packaging should pass feature configuration values to `rustc` without stripping quotes or adding backslashes.

Windows graph analysis should no longer fail when mise shims appear before Rustup executables on `PATH`. Non-Windows host lookup is unchanged.

Crates with build scripts now avoid the Git Bash wrapper path when invoking native Windows `rustc`.

Prelude-only changes will now rebuild `once-frontend` reliably when Cargo restores cached outputs.

The pull request workflow gets one additional Windows job when Rust files change. The job builds like release but does not package, upload, publish, or pass a GitHub token to the build command.

## Validation

- `mise exec -- cargo test -p once-frontend --test prelude`
- `mise exec -- cargo test -p once-frontend windows_host_which_candidates`
- `mise exec -- cargo test -p once-cli rustc_response_args_keep_arguments_verbatim`
- `mise exec -- cargo test -p once-cli materialize_declared_arg_files_writes_rustc_response_args`
- `mise exec -- cargo test -p once-cli`
- `mise exec -- cargo test -p once-frontend`
- `mise exec -- cargo fmt --all -- --check`
- `git diff --check`
- `mise exec -- cargo clippy -p once-cli -p once-frontend --all-targets -- -D warnings`
- `mise exec -- cargo clippy -p once-frontend --all-targets -- -D warnings`
- `mise exec -- cargo build --locked --release --package once-cli`
- `ruby -e 'require "yaml"; YAML.load_file(".github/workflows/once.yml"); puts "yaml ok"'`

### CODE REVIEW

File: .github/workflows/once.yml [hunk 1]
```diff
@@ -226,6 +226,43 @@
...
+        shell: pwsh
+        run: '"HOME=$env:USERPROFILE" | Out-File -FilePath $env:GITHUB_ENV -Encoding utf8 -Append'
+      - uses: jdx/mise-action@v4
+        with:
+          experimental: true
+          cache: true
+          install_args: "rust jq"
+      - name: Add Rust target
+        run: rustup target add x86_64-pc-windows-msvc
+      - uses: Swatinem/rust-cache@v2
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply .github/workflows/once.yml]: Pinned the new Windows release graph build cache step to the same Swatinem rust-cache commit used by the release workflow, so the added step no longer depends on the mutable v2 tag.

File: .github/workflows/once.yml [hunk 2]
```diff
@@ -226,6 +226,45 @@ jobs:
...
+      - name: Check Rust toolchain
+        run: mise exec -- rustc -vV
+      - uses: Swatinem/rust-cache@v2
+        with:
+          shared-key: release-package-windows
+          cache-on-failure: true
+          cache-bin: false
+      - name: Package Windows release archive
+        run: mise run release:package --target x86_64-pc-windows-msvc --version 0.0.0
+        env:
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply .github/workflows/once.yml]: Changed the check to build the Windows release graph target directly instead of running release:package. It no longer passes GH_TOKEN and does not create package archives or release assets.

File: .github/workflows/once.yml [hunk 3]
```diff
@@ -45,7 +45,7 @@ jobs:
           rust_ci=false
           while IFS= read -r path; do
             case "$path" in
-              Cargo.toml|Cargo.lock|mise.toml|once.toml|crates/*|spec/*|mise/tasks/*|.cargo/*)
+              Cargo.toml|Cargo.lock|mise.toml|once.toml|crates/*|spec/*|mise/tasks/*|.cargo/*|.github/workflows/*)
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply .github/workflows/once.yml]: Reverted the workflow path filter expansion. This new check still runs for this branch because the Rust graph files changed, but workflow-only edits will not fan out into the full Rust workflow.

File: .github/workflows/once.yml [hunk 4]
```diff
@@ -226,6 +226,45 @@ jobs:
...
+      - name: Add Rust target
+        run: rustup target add x86_64-pc-windows-msvc
+      - name: Check Rust toolchain
+        run: mise exec -- rustc -vV
+      - uses: Swatinem/rust-cache@v2
+        with:
+          shared-key: release-package-windows
+          cache-on-failure: true
+          cache-bin: false
+      - name: Package Windows release archive
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply .github/workflows/once.yml]: Removed the release:package invocation. The check now writes release manifests and builds the Windows release graph target, so 0.0.0 is only a local build metadata input and no package or release artifact is produced.

File: crates/once-cli/src/commands/graph/analysis/actions.rs [hunk 5]
```diff
@@ -916,60 +895,23 @@ mod tests {
...
     }
 
     #[test]
-    fn rustc_response_args_keep_quotes_for_non_windows_hosts() {
+    fn rustc_response_args_keep_arguments_verbatim() {
         let arg_file = DeclaredArgFile {
             path: ".once/out/rust/rustc-features.rsp".to_string(),
             format: DeclaredArgFileFormat::RustcResponse,
             args: Vec::new(),
         };
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply crates/once-cli/src/commands/graph/analysis/actions.rs]: Added edge-case coverage for the rustc response-file verbatim contract in crates/once-cli/src/commands/graph/analysis/actions.rs. The test now covers empty arguments, spaces, backslashes, tabs, single quotes, and double quotes. Validation: `mise exec -- cargo test -p once-cli rustc_response_args_keep_arguments_verbatim` and `mise exec -- cargo fmt --all -- --check`.

File: crates/once-cli/src/commands/graph/analysis/actions.rs [hunk 6]
```diff
@@ -916,62 +895,32 @@ mod tests {
...
-        );
+        let cases = [
+            "",
+            "argument with spaces",
+            r"C:\Program Files\Rust\lib",
+            "tab\tseparated",
+            "feature='alloc'",
+            "feature=\"alloc\"",
+        ];
+
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply crates/once-cli/src/commands/graph/analysis/actions.rs:920]: Added a contextual assertion message to the rustc response argument loop, so any failure reports the exact input with debug formatting. Validation: `mise exec -- cargo test -p once-cli rustc_response_args_keep_arguments_verbatim` and `mise exec -- cargo fmt --all -- --check`.

File: crates/once-frontend/build.rs [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    let manifest_dir = PathBuf::from(
+        env::var_os("CARGO_MANIFEST_DIR").expect("CARGO_MANIFEST_DIR is set for build scripts"),
+    );
+    emit_rerun_if_changed(&manifest_dir.join("prelude"))
+}
+
+fn emit_rerun_if_changed(path: &Path) -> io::Result<()> {
+    println!("cargo:rerun-if-changed={}", path.display());
+
+    if path.is_dir() {
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply crates/once-frontend/build.rs]: Updated the prelude input tracker to use `symlink_metadata` before recursing and to sort directory entries before emitting `cargo:rerun-if-changed` lines. That keeps the build dependency output deterministic and avoids following symlinks outside the embedded prelude tree. Validation: `mise exec -- cargo fmt --all -- --check`, `mise exec -- cargo test -p once-frontend --test prelude prelude_cargo_metadata_windows_features_keep_response_file_cfgs_literal`, and `mise exec -- cargo build --locked --release --package once-cli`.

File: crates/once-frontend/build.rs [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    let manifest_dir = PathBuf::from(
+        env::var_os("CARGO_MANIFEST_DIR").expect("CARGO_MANIFEST_DIR is set for build scripts"),
+    );
+    emit_rerun_if_changed(&manifest_dir.join("prelude"))
+}
+
+fn emit_rerun_if_changed(path: &Path) -> io::Result<()> {
+    println!("cargo:rerun-if-changed={}", path.display());
+
+    if !fs::symlink_metadata(path)?.is_dir() {
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply crates/once-frontend/build.rs]: Adjusted the prelude tracker to follow directory symlinks through `metadata`, then bound traversal with canonical paths and a visited set. That supports a symlinked prelude root, avoids cycles, and does not recurse into links whose canonical path leaves the prelude root. Validation: `mise exec -- cargo fmt --all -- --check`, `mise exec -- cargo clippy -p once-frontend --all-targets -- -D warnings`, and `mise exec -- cargo build --locked --release --package once-cli`.

File: crates/once-frontend/build.rs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    let mut visited = BTreeSet::new();
+    emit_rerun_if_changed(&prelude, &root, &mut visited)
+}
+
+fn emit_rerun_if_changed(
+    path: &Path,
+    root: &Path,
+    visited: &mut BTreeSet<PathBuf>,
+) -> io::Result<()> {
+    println!("cargo:rerun-if-changed={}", path.display());
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply crates/once-frontend/build.rs:44]: Fixed by moving canonicalization and the prelude-root boundary check ahead of the `cargo:rerun-if-changed` emission in `crates/once-frontend/build.rs`.

Broken symlinks and symlinks that resolve outside `prelude` now return without being registered. In-tree directory symlinks still recurse through the canonical visited set, so legitimate symlinked prelude layouts continue to rebuild when nested files change.

Validation:
- `mise exec -- cargo clippy -p once-frontend --all-targets -- -D warnings`
- `mise exec -- cargo test -p once-frontend --test prelude prelude_cargo_metadata_windows_features_keep_response_file_cfgs_literal`
- `mise exec -- cargo test -p once-frontend --test prelude prelude_rust_windows_feature_cfgs_use_response_file`
- `mise exec -- cargo fmt --all -- --check`
- `git diff --check`

File: .github/workflows/once.yml [hunk 10]
```diff
@@ -226,6 +226,81 @@ jobs:
...
+          shared-key: release-graph-build-windows
+          cache-on-failure: true
+          cache-bin: false
+      - name: Build Windows release binary with Once
+        run: |
+          set -euo pipefail
+
+          probe_dir="$(mktemp -d)"
+          printf 'fn main() {}\n' > "${probe_dir}/probe.rs"
+          printf '%s\n' '--cfg=feature="default"' > "${probe_dir}/rustc-features.rsp"
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply .github/workflows/once.yml]: Removed the hand-written rustc probe. The Windows job now relies on the actual Once release graph build to materialize and consume `.once/out/.../rustc-features.rsp` through `DeclaredArgFileFormat::RustcResponse`, and it only dumps the generated response files when that build fails.

Validation:
- `ruby -e 'require "yaml"; YAML.load_file(".github/workflows/once.yml"); puts "yaml ok"'`
- `mise exec -- cargo fmt --all -- --check`
- `mise exec -- cargo clippy -p once-frontend --all-targets -- -D warnings`
- `git diff --check`

File: crates/once-frontend/build.rs [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+    let mut entries = fs::read_dir(path)?.collect::<io::Result<Vec<_>>>()?;
+    entries.sort_by_key(std::fs::DirEntry::path);
+    for entry in entries {
+        emit_rerun_if_changed(&entry.path(), root, visited)?;
+    }
+
+    Ok(())
+}
+
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply crates/once-frontend/build.rs]: Simplified `crates/once-frontend/build.rs` so `path.canonicalize()` errors now propagate. Broken symlinks fail the build instead of being treated as absent, while symlinks that resolve outside the prelude root are still skipped before any `cargo:rerun-if-changed` line is emitted.

Validation:
- `mise exec -- cargo fmt --all -- --check`
- `mise exec -- cargo clippy -p once-frontend --all-targets -- -D warnings`
- `git diff --check`

File: crates/once-frontend/src/analysis/tests.rs [hunk 12]
```diff
@@ -98,6 +98,37 @@ fn host_file_contains_matches_binary_host_files() {
     assert!(contains);
 }
 
+#[test]
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply crates/once-frontend/src/analysis/tests.rs:103]: Added a Windows-only test that calls the production `which_candidate_names` function with a controlled `PATHEXT`, guarded by a mutex and restored afterward. The deterministic helper tests stay in place for non-Windows hosts.

Validation:
- `mise exec -- cargo test -p once-frontend windows_host_which_candidates`
- `mise exec -- cargo clippy -p once-frontend --all-targets -- -D warnings`
- `mise exec -- cargo fmt --all -- --check`

File: crates/once-frontend/src/analysis/tests.rs [hunk 13]
```diff
@@ -98,6 +100,64 @@ fn host_file_contains_matches_binary_host_files() {
...
+        vec!["rustc.exe".to_string()]
+    );
+}
+
+#[test]
+fn unix_host_which_candidates_keep_name_verbatim() {
+    assert_eq!(
+        which_candidate_names_for("rustc", false, Some(".COM;.EXE;.CMD")),
+        vec!["rustc".to_string()]
+    );
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply crates/once-frontend/src/analysis/tests.rs:131]: Fixed by changing the Windows-only production-path test to read the live `PATHEXT` without mutating it. It now compares `which_candidate_names("rustc")` with the deterministic helper using the current environment value, so it still exercises the wrapper without changing process-global state.

Validation:
- `mise exec -- cargo test -p once-frontend windows_host_which_candidates`
- `mise exec -- cargo clippy -p once-frontend --all-targets -- -D warnings`
- `mise exec -- cargo fmt --all -- --check`
- `git diff --check`

File: crates/once-frontend/build.rs [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    let mut visited = BTreeSet::new();
+    emit_rerun_if_changed(&prelude, &root, &mut visited)
+}
+
+fn emit_rerun_if_changed(
+    path: &Path,
+    root: &Path,
+    visited: &mut BTreeSet<PathBuf>,
+) -> io::Result<()> {
+    let canonical = path.canonicalize()?;
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply crates/once-frontend/build.rs]: Fixed by restoring fallible handling for missing entries while keeping the boundary check before normal dependency emission. `build.rs` now uses `symlink_metadata` first, skips transient missing paths, registers unresolved symlink paths themselves, and only emits regular entries after the canonical path is confirmed inside the prelude root.

Validation:
- `mise exec -- cargo clippy -p once-frontend --all-targets -- -D warnings`
- `mise exec -- cargo fmt --all -- --check`
- `git diff --check`

File: .github/workflows/once.yml [hunk 15]
```diff
@@ -226,6 +226,73 @@ jobs:
...
+      - uses: Swatinem/rust-cache@e18b497796c12c097a38f9edb9d0641fb99eee32
+        with:
+          shared-key: release-graph-build-windows
+          cache-on-failure: true
+          cache-bin: false
+      - name: Build Windows release binary with Once
+        run: |
+          set -euo pipefail
+
+          dump_rustc_feature_response_files() {
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply .github/workflows/once.yml]: Restored the focused Windows rustc response-file probe before the full release graph build. The probe covers the compiler syntax directly, and the graph build still validates Once's generated `rustc-features.rsp` files through the actual release-like path.

Validation:
- `ruby -e 'require "yaml"; YAML.load_file(".github/workflows/once.yml"); puts "yaml ok"'`
- `mise exec -- cargo fmt --all -- --check`
- `git diff --check`

File: .github/workflows/once.yml [hunk 16]
```diff
@@ -226,6 +226,81 @@ jobs:
...
+          cache-on-failure: true
+          cache-bin: false
+      - name: Build Windows release binary with Once
+        run: |
+          set -euo pipefail
+
+          probe_dir="$(mktemp -d)"
+          printf 'fn main() {}\n' > "${probe_dir}/probe.rs"
+          printf '%s\n' '--cfg=feature="default"' > "${probe_dir}/rustc-features.rsp"
+          rustc "@${probe_dir}/rustc-features.rsp" "${probe_dir}/probe.rs" \
```
[the remark this answers is not in the document]
Comment (pepicrft) [reply .github/workflows/once.yml]: Removed the standalone rustc probe entirely and simplified the Windows release graph job to validate only the release-like Once graph build path. The workflow now stays on the mise-managed toolchain path and avoids maintaining a second compiler invocation in YAML.

Validation:
- `ruby -e 'require "yaml"; YAML.load_file(".github/workflows/once.yml"); puts "yaml ok"'`
- `git diff --check`
